tests/unit/test_oauth_provider.py

```diff
@@ -0,0 +1,200 @@
+        "issuer_url": "https://example.oidc.com",
+        "jwks_endpoint": "https://example.oidc.com/.well-known/jwks.json",
+        "scope": "openid profile email phone",
+        "token_endpoint": "https://example.oidc.com/oauth2/token",
+        "userinfo_endpoint": "https://example.oidc.com/userinfo",
+        "client_id": CLIENT_ID,
+    }
+
+
+def test_client_changed(harness):
+    relation_id = harness.add_relation("oauth", "requirer")
+    harness.add_relation_unit(relation_id, "requirer/0")
+    harness.update_relation_data(
+        relation_id,
+        "requirer",
+        {
+            "redirect_uri": "https://oidc-client.com/callback",
+            "scope": "openid email",
+            "grant_types": '["authorization_code"]',
+            "audience": "[]",
+            "token_endpoint_auth_method": "client_secret_basic",
+        },
+    )
+    assert len(harness.charm.events) == 1
+    assert isinstance(harness.charm.events[0], ClientCreatedEvent)
+
+    redirect_uri = "https://oidc-client.com/callback2"
+    harness.update_relation_data(
+        relation_id,
+        "requirer",
+        {
+            "redirect_uri": redirect_uri,
+            "scope": "openid email",
+            "grant_types": '["authorization_code"]',
+            "audience": "[]",
+            "token_endpoint_auth_method": "client_secret_basic",
+        },
+    )
+
+    assert len(harness.charm.events) == 2
+    assert isinstance(harness.charm.events[1], ClientChangedEvent)
+    assert harness.charm.events[1].redirect_uri == redirect_uri
+
+
+def test_on_client_config_deleted_event_emitted(harness):
+    relation_id = harness.add_relation("oauth", "requirer")
+    harness.add_relation_unit(relation_id, "requirer/0")
+    harness.update_relation_data(
+        relation_id,
+        "requirer",
+        {
+            "redirect_uri": "https://oidc-client.com/callback",
+            "scope": "openid email",
+            "grant_types": '["authorization_code"]',
+            "audience": "[]",
+            "token_endpoint_auth_method": "client_secret_basic",
+        },
+    )
+    assert len(harness.charm.events) == 1
+    assert isinstance(harness.charm.events[0], ClientCreatedEvent)
+
+    harness.remove_relation(relation_id)
+
+    assert len(harness.charm.events) == 2
+    assert isinstance(harness.charm.events[1], ClientDeletedEvent)
+
+
+def test_on_client_config_deleted_secret_removed(harness):
+    relation_id = harness.add_relation("oauth", "requirer")
+    harness.add_relation_unit(relation_id, "requirer/0")
+    harness.update_relation_data(
+        relation_id,
+        "requirer",
+        {
+            "redirect_uri": "https://oidc-client.com/callback",
+            "scope": "openid email",
+            "grant_types": '["authorization_code"]',
+            "audience": "[]",
+            "token_endpoint_auth_method": "client_secret_basic",
+        },
+    )
+    assert len(harness.charm.events) == 1
```

Comment:
Your test should focus on validating 1 behavior. I'd remove these 2 asserts

Reply:
If those asserts were to fail that means that the rest of the test will not work, as the client will not be created thus making it impossible to remove the client_secret. They are there so that the dev will get a clearer message in case something goes wrong.

Reply:
Sure but the goal of the test is to validate 1 behavior and this assert validates "something else". If you really want to validate this behavior you may have a test specifically for it but I'd discourage testing for the same thing in multiple tests.